### PR TITLE
osx: set the default thread stack size to RLIMIT_STACK

### PR DESCRIPTION
**Pull request by saghul**
Fixes: https://github.com/libuv/libuv/issues/669

R=@bnoordhuis

I'll try to write a test later today.

### CODE REVIEW

File: src/unix/thread.c [hunk 1]
```diff
@@ -63,7 +69,27 @@ int uv_thread_create(uv_thread_t *tid, void (*entry)(void *arg), void *arg) {
...
+   */
+#if defined(__APPLE__)
+  if (getrlimit(RLIMIT_STACK, &lim))
+    abort();
+
+  attr = &attr_storage;
+  if (pthread_attr_init(attr))
+    abort();
+
+  if (pthread_attr_setstacksize(attr, lim.rlim_cur))
```
Comment (bnoordhuis) [inline src/unix/thread.c]: You should guard against `lim.rlim_cur == RLIM_INFINITY` here.  You should probably also do `max(lim.rlim_cur, PTHREAD_STACK_MIN)`.

Comment (saper) [inline src/unix/thread.c]: Do we need to `abort()` on every failure? In that case I'd just go ahead with the defaults, maybe.

Comment (saper) [inline src/unix/thread.c]: Also I was wondering if we really need to  `#ifdef __APPLE__` it.

**Comment by saper** [comment]
@xzyfer @58bits can you try to apply this change to your node engine source, recompile and see if that alleviates the problem we have in https://github.com/sass/node-sass/issues/1289 ?

File: src/unix/thread.c (hunk 1 above)
Comment (bnoordhuis) [inline src/unix/thread.c]: I think aborting is the proper thing to do here.  These functions are not expected to fail.

Comment (saper) [inline src/unix/thread.c]: Yes, you're right. Failing right away if the sky is falling is a better policy.

Comment (saghul) [inline src/unix/thread.c]: > You should guard against lim.rlim_cur == RLIM_INFINITY here.

Is that remotely possible?

> You should probably also do max(lim.rlim_cur, PTHREAD_STACK_MIN).

Will do.

**Comment by saghul** [comment]
@bnoordhuis updated, can you PTAL?

**Comment by 58bits** [comment]
Just tried compiling stable node source with the updated `src/unix/thread.c` and got the following error...
```
../deps/uv/src/unix/thread.c:83:25: error: use of undeclared identifier 'RLIMIT_INFINITY'
    if (lim.rlim_cur == RLIMIT_INFINITY) {
```
`RLIMIT_STACK` is defined in `sys/resource.h`

Where is `RLIMIT_INFINITY`?

**Comment by saghul** [comment]
@58bits sorry, you caught the code in flux, it's fixed now and I added a test for it.

@bnoordhuis can you PTAL now?

**Comment by 58bits** [comment]
Ok thanks @saghul! I'll try again :)

**Comment by 58bits** [comment]
@saper @xzyfer @Snugug this works! I've just compiled Node stable v5.3.0 with @saghul's patch and node-sass compiles singularity style sheets in debug mode fine. w00t!

**Comment by saghul** [comment]
@58bits sweet, thanks for confirming!

File: src/unix/thread.c [hunk 2]
```diff
@@ -63,7 +72,32 @@ int uv_thread_create(uv_thread_t *tid, void (*entry)(void *arg), void *arg) {
...
+  if (getrlimit(RLIMIT_STACK, &lim))
+    abort();
+
+  attr = &attr_storage;
+  if (pthread_attr_init(attr))
+    abort();
+
+  if (lim.rlim_cur == RLIM_INFINITY) {
+    pthread_attr_destroy(attr);
+    attr = NULL;
```
Comment (bnoordhuis) [inline src/unix/thread.c]: Not wrong but you can just let it fall through and clean up after pthread_create().

In retrospect, you can do the same if `lim.rlim_cur < PTHREAD_STACK_MIN`, i.e., simply don't call pthread_attr_setstacksize() and create the thread with default attributes.

File: src/unix/thread.c (hunk 1 above)
Comment (bnoordhuis) [inline src/unix/thread.c]: > Is that remotely possible?

Yes, `ulimit -s unlimited` is legal (although you probably need to be root.)

**Comment by bnoordhuis** [comment]
LGTM with a suggestion.

File: src/unix/thread.c (hunk 2 above)
Comment (saghul) [inline src/unix/thread.c]: Ah, yeah, lets do that.

**Comment by saghul** [comment]
@bnoordhuis pushed a fix simplifying the code as suggested.

File: src/unix/thread.c [hunk 3]
```diff
@@ -63,7 +69,30 @@ int uv_thread_create(uv_thread_t *tid, void (*entry)(void *arg), void *arg) {
...
+#if defined(__APPLE__)
+  if (getrlimit(RLIMIT_STACK, &lim))
+    abort();
+
+  attr = &attr_storage;
+  if (pthread_attr_init(attr))
+    abort();
+
+  if (lim.rlim_cur != RLIM_INFINITY &&
+      lim.rlim_cur > PTHREAD_STACK_MIN) {
```
Comment (bnoordhuis) [inline src/unix/thread.c]: `>=`?

**Comment by bnoordhuis** [comment]
LGTM with a nit.

**Comment by saghul** [comment]
Addressed the nit and merged, thanks as always Ben!

**Comment by 58bits** [comment]
Hi All, was just checking if this had made it into node/deps/uv/src/unix/ - and see that the patch is in the v1.x branch, but not yet in a tagged release. Is there an ETA for the next libuv release?

**Comment by saghul** [comment]
No ETA yet.

**Comment by 58bits** [comment]
Okay thanks @saghul - understood.

**Pull request merged**